accounts-db/benches/read_only_accounts_cache.rs

```diff
@@ -0,0 +1,287 @@
+                            let mut rng = SmallRng::seed_from_u64(100_u64.saturating_add(i as u64));
+                            while !stop_threads.load(Ordering::Relaxed) {
+                                let (pubkey, account) = accounts.choose(&mut rng).unwrap();
+                                cache.store(*pubkey, slot, account.clone());
+                            }
+                        }
+                    })
+                    .unwrap()
+            })
+            .collect::<Vec<_>>();
+
+        group.bench_function(
+            BenchmarkId::new("read_only_accounts_cache_store", num_readers_writers),
+            |b| {
+                b.iter_custom(|iters| {
+                    let mut total_time = Duration::new(0, 0);
+
+                    for (pubkey, account) in accounts.iter().cycle().take(iters as usize) {
+                        // Measure only stores.
+                        let start = Instant::now();
+                        cache.store(*pubkey, slot, account.clone());
+                        total_time = total_time.saturating_add(start.elapsed());
+                    }
+                    total_time
+                })
+            },
+        );
+        group.bench_function(
+            BenchmarkId::new("read_only_accounts_cache_load", num_readers_writers),
+            |b| {
+                b.iter_custom(|iters| {
+                    let start = Instant::now();
+                    for (pubkey, _) in accounts.iter().cycle().take(iters as usize) {
+                        test::black_box(cache.load(*pubkey, slot));
+                    }
+
+                    start.elapsed()
+                })
+            },
+        );
+
+        stop_threads.store(true, Ordering::Relaxed);
+        for reader_handle in reader_handles {
+            reader_handle.join().unwrap();
+        }
+        for writer_handle in writer_handles {
+            writer_handle.join().unwrap();
+        }
+    }
+}
+
+/// Benchmarks the read-only cache eviction mechanism. It does so by performing
+/// multithreaded reads and writes on a full cache. Each write triggers
+/// eviction. Background reads add more contention.
+fn bench_read_only_accounts_cache_eviction(c: &mut Criterion) {
+    // Prepare initial accounts, enough of the to fill up the cache.
+    let accounts: Vec<_> = utils::accounts_with_size_limit(
+        255,
+        AccountsDb::DEFAULT_MAX_READ_ONLY_CACHE_DATA_SIZE_HI * 10,
+    )
+    .collect();
+    let pubkeys: Vec<_> = accounts
+        .iter()
+        .map(|(pubkey, _)| pubkey.to_owned())
+        .collect();
+
+    let mut group = c.benchmark_group("read_only_accounts_cache_eviction");
+
+    for num_readers_writers in NUM_READERS_WRITERS {
+        let cache = Arc::new(ReadOnlyAccountsCache::new(
+            AccountsDb::DEFAULT_MAX_READ_ONLY_CACHE_DATA_SIZE_LO,
+            AccountsDb::DEFAULT_MAX_READ_ONLY_CACHE_DATA_SIZE_HI,
```

Comment:
I don't think that the goal of the bench should be to literally cause an evict() call at every store() tho? (might need to clarify this somewhere). Rather that every store is beyond the max size, so conceptually triggers an eviction. _When_ the eviction happens exactly seems like an impl detail, I think we want to use _LO and _HI to model exactly what happens in the validator

Reply:
I agree with @alessandrod here. I wanted the evictor to behave as it would on MNB.

Reply:
Fair, I can add the "max eviction" case. I do agree that the "no evictions" and "somewhere in the middle" are the cases I covered now.

> My concern with "somewhere in the middle" is that runtime duration from iteration to iteration may/will be very different if there is vs isn't an eviction.

That's true. But I think that because of high number of iterations (between dozens and hundreds of thousands for each sample) and benchmark taking an average of them per sample, I think eviction still gets reflected in the average time we end up seeing at the end.

Alternatively, we could measure eviction of N elements during one criterion iteration by doing something like:

```rust
        group.bench_function(
            BenchmarkId::new(
                "read_only_accounts_cache_eviction_store",
                num_readers_writers,
            ),
            |b| {
                b.iter_custom(|iters| {
                    let mut total = Duration::new(0, 0);
                    for _ in 0..iters {
                        let limit = 1000; // whatever is good for filling up from _LO to _HI and adding some more stuff on top
                        let accounts = utils::accounts(0, DATA_SIZES, WEIGHTS).take(limit);

                        let start = Instant::now();
                        for (pubkey, account) in accounts {
                            cache.store(pubkey, slot, account);
                        }

                        total = total.saturating_add(start.elapsed());
                    }

                    total
                })
            },
        );
```

Then we would make sure that each criterion iteration represents a full eviction case. In fact, I think the very first version of benchmark I pushed here was close to that. The downside will be that the benchmark will end up with way longer execution time.

Reply:
I added the max case, PTAL